### PR TITLE
Initial Store & Forward Implementation

### PR DESCRIPTION
**Pull request by devonh**

### CODE REVIEW

File: federationtypes.go [hunk 1]
```diff
@@ -114,6 +114,13 @@ type RespSend struct {
 	PDUs map[string]PDUResult `json:"pdus"`
 }
 
+// A RespGetRelayTransaction is the content of a response to GET /_matrix/federation/v1/relay_txn/{userID}/
+type RespGetRelayTransaction struct {
```
Comment (kegsay) [inline federationtypes.go:118]: Put all the relay stuff in `relaytypes.go` please.

File: request.go [hunk 2]
```diff
@@ -212,17 +212,21 @@ func VerifyHTTPRequest(
...
 	}
-	if request.fields.Destination != "" {
-		switch {
-		case isLocalServerName != nil && !isLocalServerName(request.fields.Destination):
-			fallthrough
-		case isLocalServerName == nil && destination != request.fields.Destination:
-			message := fmt.Sprintf("Unrecognised server name %q for Destination", request.fields.Destination)
-			util.GetLogger(req.Context()).Warn(message)
-			return nil, util.MessageResponse(400, message)
+	if destination != "" {
```
Comment (kegsay) [inline request.go:215]: Why? Does this risk breaking validation?

File: request.go [hunk 3]
```diff
@@ -212,17 +212,21 @@ func VerifyHTTPRequest(
...
+		if request.fields.Destination != "" {
+			switch {
+			case isLocalServerName != nil && !isLocalServerName(request.fields.Destination):
+				fallthrough
+			case isLocalServerName == nil && destination != request.fields.Destination:
+				println(destination)
+				println(request.fields.Destination)
```
Comment (kegsay) [inline request.go:222]: Debug logging :)

File: user_test.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	}
+	if userID.Domain() != gomatrixserverlib.ServerName(domain) {
+		t.Fatalf("Domain - Expected: %s Actual: %s ", domain, userID.Domain())
+	}
+	if userID.Raw() != raw {
+		t.Fatalf("Raw - Expected: %s Actual: %s ", raw, userID.Raw())
+	}
+}
+
+func TestDomainWithPortSucceeds(t *testing.T) {
```
Comment (kegsay) [inline user_test.go:83]: Table-driven tests would be preferable here, for all these tests. 

https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

**Review by kegsay — APPROVED** [review]
Otherwise LGTM

File: request.go (hunk 2 above)
Comment (devonh) [reply request.go:215]: huh. I totally forgot about this.
it seems unnecessary now so I'll just rip it out.

**Pull request merged**